src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -784,42 +785,69 @@
   "Delete paths in dropped-pids that are no longer mentioned
-  in other factsets."
+   in other factsets."
   [factset-id dropped-pids]
   (when-let [dropped-pids (seq dropped-pids)]
     (let [in-pids (jdbc/in-clause dropped-pids)]
-      (sql/do-prepared
-       (format
-        "DELETE FROM fact_paths fp
-           WHERE fp.id %s
-             AND NOT EXISTS (SELECT 1 FROM facts f
-                               WHERE f.fact_path_id %s
-                                 AND f.fact_path_id = fp.id
-                                 AND f.factset_id <> ?)"
-        in-pids in-pids)
-       (concat dropped-pids dropped-pids [factset-id])))))
+      (if (sutils/postgres?)
+        (sql/do-prepared
+          "DELETE FROM fact_paths fp
+           WHERE fp.id = ANY (?)
+           AND NOT EXISTS (SELECT 1 FROM facts f
+             WHERE f.fact_path_id = ANY (?)
+             AND f.fact_path_id = fp.id
+             AND f.factset_id <> ?)"
+          (flatten [(sutils/array-to-param "bigint" Long dropped-pids)
+                    (sutils/array-to-param "bigint" Long dropped-pids)
+                    factset-id]))
+        (sql/do-prepared
+          (format
+            "DELETE FROM fact_paths fp
+               WHERE fp.id %s
+                 AND NOT EXISTS (SELECT 1 FROM facts f
+                                   WHERE f.fact_path_id %s
+                                     AND f.fact_path_id = fp.id
+                                     AND f.factset_id <> ?)"
+            in-pids in-pids)
+          (concat dropped-pids dropped-pids [factset-id]))))))
 
 (defn-validated delete-pending-value-id-orphans!
   "Delete values in removed-pid-vid-pairs that are no longer mentioned
-  in facts."
+   in facts."
   [factset-id removed-pid-vid-pairs]
   (when-let [removed-pid-vid-pairs (seq removed-pid-vid-pairs)]
     (let [vids (map second removed-pid-vid-pairs)
           rm-facts (map cons (repeat factset-id) removed-pid-vid-pairs)
           in-vids (jdbc/in-clause vids)
           in-rm-facts (jdbc/in-clause-multi rm-facts 3)]
-      (sql/do-prepared
-       (format
-        "DELETE FROM fact_values fv
-           WHERE fv.id %s
+      (if (sutils/postgres?)
+        (sql/do-prepared
+          (format
+            "DELETE FROM fact_values fv
+             WHERE fv.id = ANY(?)
              AND NOT EXISTS (SELECT 1 FROM facts f
-                               WHERE f.fact_value_id %s
-                                 AND f.fact_value_id = fv.id
-                                 AND (f.factset_id,
-                                      f.fact_path_id,
-                                      f.fact_value_id) NOT %s)"
-        in-vids in-vids in-rm-facts)
-       (flatten [vids vids rm-facts])))))
+               WHERE f.fact_value_id = ANY(?)
+               AND f.fact_value_id = fv.id
+               AND (f.factset_id,
+                 f.fact_path_id,
+                 f.fact_value_id)
+               NOT %s)"
```

Comment:
As we talked about IRL, it would be great if this was passed as an array parameter too